// material-ui-repeatable/Repeatable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _FlatButton = require('material-ui/FlatButton');

var _FlatButton2 = _interopRequireDefault(_FlatButton);

var _FontIcon = require('material-ui/FontIcon');

var _FontIcon2 = _interopRequireDefault(_FontIcon);

var _fontAwesome = require('font-awesome/css/font-awesome.css');

var _fontAwesome2 = _interopRequireDefault(_fontAwesome);

var _Repeatable = require('./Repeatable.css');

var _Repeatable2 = _interopRequireDefault(_Repeatable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var plus = _react2.default.createElement(_FontIcon2.default, { className: _fontAwesome2.default.fa + ' ' + _fontAwesome2.default['fa-plus'] });
var minus = _react2.default.createElement(_FontIcon2.default, { className: _fontAwesome2.default.fa + ' ' + _fontAwesome2.default['fa-minus'] });

var Repeatable = function (_Component) {
    _inherits(Repeatable, _Component);

    function Repeatable() {
        _classCallCheck(this, Repeatable);

        var _this = _possibleConstructorReturn(this, (Repeatable.__proto__ || Object.getPrototypeOf(Repeatable)).call(this));

        _this.removeItem = function (index) {
            return function () {
                var items = _extends({}, _this.state.items);
                delete items[index];
                _this.setState({ items: items });
                _this.props.onRemove(index);
            };
        };

        _this.createItems = function () {
            var num = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;

            var items = _extends({}, _this.state.items);
            for (var index = 0; index < num; index++) {
                var key = _this.counter++;
                items[key] = true;
            }
            _this.setState({ items: items });
            return _this.counter;
        };

        _this.addItem = function () {
            return _this.props.onAdd(_this.createItems(1));
        };

        _this.counter = 0;
        _this.state = {
            items: {}
        };
        return _this;
    }

    _createClass(Repeatable, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.createItems(this.props.initialCopies);
        }
    }, {
        key: 'renderItems',
        value: function renderItems() {
            var _this2 = this;

            return (0, _lodash.map)(this.state.items, function (_, index) {
                return _react2.default.createElement(
                    'div',
                    { key: 'repeatable-item-' + index, className: _Repeatable2.default.Row },
                    _react2.default.createElement(
                        'div',
                        { className: _Repeatable2.default.Item },
                        _this2.props.childrenRenderer(index)
                    ),
                    _react2.default.createElement(
                        'div',
                        {
                            className: _Repeatable2.default.ItemControls,
                            style: { display: index < _this2.props.requiredCopies ? 'none' : '' } },
                        _react2.default.createElement(_FlatButton2.default, { icon: minus, onClick: _this2.removeItem(index) })
                    )
                );
            });
        }
    }, {
        key: 'render',
        value: function render() {
            return _react2.default.createElement(
                'div',
                { style: this.props.style },
                this.renderItems(),
                _react2.default.createElement(
                    'div',
                    { className: _Repeatable2.default.Controls },
                    _react2.default.createElement(_FlatButton2.default, { icon: plus, onClick: this.addItem })
                )
            );
        }
    }]);

    return Repeatable;
}(_react.Component);

Repeatable.defaultProps = {
    initialCopies: 1,
    requiredCopies: 1,
    childrenRenderer: function childrenRenderer(index) {
        return [];
    },
    onAdd: function onAdd(index) {
        return undefined;
    },
    onRemove: function onRemove(index) {
        return undefined;
    },
    style: {}
};
exports.default = Repeatable;
